In [18]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.models import load_model

In [19]:
word_to_index= imdb.get_word_index()
reverse_word_index= {value: key for key, value in word_to_index.items()}

In [20]:
model = load_model('pickle\simple_rnn_best_model.h5')
model.summary()

<>:1: SyntaxWarning: invalid escape sequence '\s'
<>:1: SyntaxWarning: invalid escape sequence '\s'
C:\Users\Rondon\AppData\Local\Temp\ipykernel_30164\717625024.py:1: SyntaxWarning: invalid escape sequence '\s'
  model = load_model('pickle\simple_rnn_best_model.h5')


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ (None, 500, 128)       │     1,280,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn (SimpleRNN)          │ (None, 500, 128)       │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ simple_rnn_1 (SimpleRNN)        │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,345,923 (5.13 MB)

 Trainable params: 1,345,921 (5.13 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 2 (12.00 B)

In [21]:
model.get_weights()

[array([[ 0.03113127, -0.05517722,  0.00903709, ...,  0.02276902,
         -0.01742078,  0.05016988],
        [-0.00949253, -0.01643268, -0.00638177, ..., -0.03210164,
          0.01201123, -0.05587407],
        [-0.0423912 ,  0.00121362, -0.0022111 , ...,  0.0288466 ,
          0.02017921, -0.02983032],
        ...,
        [ 0.04696506,  0.0019487 , -0.043566  , ..., -0.01707223,
          0.03753803,  0.01260874],
        [ 0.01961702,  0.01274722, -0.0285542 , ..., -0.00374495,
         -0.02384978,  0.01380479],
        [ 0.03024007, -0.04493069, -0.0356141 , ..., -0.01375491,
          0.00796494,  0.04121973]], dtype=float32),
 array([[-0.07268715, -0.09826351, -0.03943864, ..., -0.01092824,
         -0.02028292,  0.01856508],
        [ 0.12449361,  0.09987377, -0.00704953, ...,  0.12413868,
         -0.01704118, -0.13683173],
        [-0.04382917, -0.16845162,  0.03657102, ...,  0.08904737,
          0.02140939,  0.13897063],
        ...,
        [-0.03789884,  0.13574205,  0.1

In [22]:
def decode_review(encoded_review):
    return ' '.join([reverse_word_index.get(i -3, '?') for i in encoded_review])

def preprocess_text(text):
    words = text.lower().split()
    encoded_review = []
    for word in words:
        idx = word_to_index.get(word, 2) + 3
        if idx >= 10000:
            idx = 9999
        encoded_review.append(idx)
    padded_review = sequence.pad_sequences([encoded_review], maxlen=500)
    return padded_review

In [23]:
def predict_sentiment(review):
    preprocessed_input = preprocess_text(review)

    pred = model.predict(preprocessed_input)

    sentiment = 'Positive' if pred[0][0] > 0.5 else 'Negative'

    return sentiment, pred[0][0]


In [38]:
# example

example_review = "Good show. THE LAST OF US was indeed a great series. After hearing many raves and good things from people that I know, I finally decided to give it a chance. Like the game, it is definitely worth your time. Well made, compelling and great characters with real chemistry made this series a fun watch, a slower build to some pretty intense scenes. While it did have periods of time that it moved more slowly, it certainly never felt boring. Pedro Pascal as Joe Miller and Bella Ramsey as Ellie Williams made this a great show, and the chemistry they develop feels natural and heart warming."
example_review_2 = "There was so much potential here. Although Zombie movies and tv shows have been overplayed in the past 20-30 years, this show has some legs under it. It got me hooked in the first episode and I was really excited for the rest.Unfortunately, even though there were some saving grace episodes or moments here and there, I ended season one feeling let down. Pedro Pascal does a phenomenal job, and is the one reason I kept watching. His character slowly develops over time and you really care about him.I know a lot of people loved her, but it was hard for me to watch Bella Ramsey. She acted a couple of parts very well (especially episode 8) but other than when she is yelling and screaming and supposed to be annoying, the times when she was supposed to be likable fell flat and I had a hard time getting through those episodes."

In [39]:
sent_1, pred_1 = predict_sentiment(example_review)
sent_2, pred_2 = predict_sentiment(example_review_2)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step


In [40]:
print(f'Review: {example_review}')
print(f'Sentiment: {sent_1}') 
print(f'Prediction score: {pred_1:.2f})')

print(f'Review: {example_review_2}')
print(f'Sentiment: {sent_2}')
print(f'Prediction score: {pred_2:.2f}')






Review: Good show. THE LAST OF US was indeed a great series. After hearing many raves and good things from people that I know, I finally decided to give it a chance. Like the game, it is definitely worth your time. Well made, compelling and great characters with real chemistry made this series a fun watch, a slower build to some pretty intense scenes. While it did have periods of time that it moved more slowly, it certainly never felt boring. Pedro Pascal as Joe Miller and Bella Ramsey as Ellie Williams made this a great show, and the chemistry they develop feels natural and heart warming.
Sentiment: Positive
Prediction score: 0.98)
Review: There was so much potential here. Although Zombie movies and tv shows have been overplayed in the past 20-30 years, this show has some legs under it. It got me hooked in the first episode and I was really excited for the rest.Unfortunately, even though there were some saving grace episodes or moments here and there, I ended season one feeling let do

To evaluate the model's performance on **real-world, unseen data**, two public reviews about _The Last of Us_ series (Season 2) were selected from the [IMDB website](https://www.imdb.com/pt/title/tt3581920/reviews/?ref_=tt_ov_ql_2). The first reviewer rated the series **9.0**, while the second gave it a **4.0**, providing a useful contrast for sentiment analysis.

The model successfully identified the sentiment in both cases. The first review was correctly classified as **positive**, with a **prediction score of 0.98**, while the second review — more critical in tone — was classified as **negative**, with a **prediction score of 0.04**. These results suggest that the model is able to generalize well to public, informal reviews and correctly interpret the **emotional tone** conveyed through natural language.